relay-server/src/actors/project.rs

```diff
@@ -630,47 +635,49 @@ impl Project {
         Response::future(future)
     }
 
-    fn fetch_state(&mut self, no_cache: bool, context: &mut Context<Self>) {
+    pub fn update_state(
+        &mut self,
+        state_result: Result<ProjectStateResponse, ()>,
+        no_cache: bool,
+        context: &mut Context<ProjectCache>,
+    ) {
+        let channel = match self.state_channel.take() {
+            Some(channel) => channel,
+            None => return,
+        };
+
+        // If the channel has `no_cache` set but we are not a `no_cache` request, we have
+        // been superseeded. Put it back and let the other request take precedence.
+        if channel.no_cache && !no_cache {
+            self.state_channel = Some(channel);
+            return;
+        }
+
+        self.state_channel = None;
+        self.state = state_result.map(|resp| resp.state).ok();
+        self.update_aggregator(context);
+
+        if let Some(ref state) = self.state {
+            relay_log::debug!("project state {} updated", self.public_key);
+            channel.send(state.clone());
+        }
+    }
+
+    fn fetch_state(&mut self, no_cache: bool, context: &mut Context<ProjectCache>) {
```

Comment:
For this changeset, this is a good and pragmatic solution. However, the dependency between the project and the project cache is circular. Instead of pushing a message into the project cache, we could avoid this by returning some form of result to the project cache, so that it triggers a fetch by itself. Through that, we should be able to remove all references to `Context` from the `Project` API.

Reply:
done